File: src/lib/carto.ts
import { getDeploymentConfig } from './deploymentConfig'

const CARTO_RASTER_HOST = /^https:\/\/(?:[a-d]\.)?basemaps\.cartocdn\.com\//i

/**
 * Add the deployment's CARTO Basemaps key to a raster tile template.
 *
 * CARTO's basemap key is a browser credential by design: it is sent as `?key=` on every
 * tile request and should be restricted to this deployment's domains in CARTO. Keeping it
 * in the runtime deployment response means it is never committed or baked into the image.
 */
export function withCartoBasemapKey(url: string): string {
  const key = getDeploymentConfig().integrations?.cartoBasemapKey?.trim()
  if (!key || !CARTO_RASTER_HOST.test(url) || /(?:[?&])key=/.test(url)) return url
  return `${url}${url.includes('?') ? '&' : '?'}key=${encodeURIComponent(key)}`
}

/** Built-in CARTO raster templates, keyed at the point a map is rendered. */
export function cartoRasterTiles(style: string, subdomains: readonly string[] = ['a', 'b', 'c', 'd']): string[] {
  return subdomains.map((s) => withCartoBasemapKey(
    `https://${s}.basemaps.cartocdn.com/${style}/{z}/{x}/{y}.png`,
  ))
}

/** Apply the runtime key to a possibly persisted/custom layer without touching other hosts. */
export function keyCartoTileTemplates<T extends { tiles?: string[]; nightTiles?: string[] }>(layer: T): T {
  const tiles = layer.tiles?.map(withCartoBasemapKey)
  const nightTiles = layer.nightTiles?.map(withCartoBasemapKey)
  return tiles === layer.tiles && nightTiles === layer.nightTiles ? layer : { ...layer, tiles, nightTiles }
}

/**
 * Strip any `key=` parameter from a tile template.
 *
 * Used on the way OUT to our own backend (the Rapport/Kroki payload names the basemap the
 * operator was looking at). The server holds the same credential and applies its own — see
 * backend `app/carto.py` — so sending ours would only copy it into a request body, the
 * application log and the on-disk tile-cache filenames.
 *
 * Host-agnostic on purpose: a key pasted into a station's own custom layer is no more ours to
 * forward than CARTO's is.
 *
 * ⚠️ Parses the query rather than pattern-matching it. `new URL()` is not an option — it
 * percent-encodes the `{z}/{x}/{y}` slots and hands back something MapLibre cannot fill — but
 * substring surgery on a URL is how `monkey=1` gets eaten, so split on the real separators and
 * drop whole parameters.
 */
export function withoutCartoBasemapKey(url: string): string {
  const hash = url.indexOf('#')
  const frag = hash < 0 ? '' : url.slice(hash)
  const base = hash < 0 ? url : url.slice(0, hash)
  const q = base.indexOf('?')
  if (q < 0) return url
  const kept = base
    .slice(q + 1)
    .split('&')
    .filter((p) => p !== '' && p !== 'key' && !p.startsWith('key='))
  return (kept.length ? `${base.slice(0, q)}?${kept.join('&')}` : base.slice(0, q)) + frag
}

File: backend/app/carto.py
"""CARTO basemap key handling for SERVER-side tile fetches (Rapport/Kroki, plan pages).

The browser half of this lives in ``src/lib/carto.ts``. The two halves answer the same
question — «does this raster template carry the deployment's CARTO key?» — for two different
fetchers, and the split matters:

* The BROWSER must carry the key, because MapLibre is what talks to the CDN. It is a public
  credential there by construction, protected by CARTO's domain restrictions rather than by
  secrecy.
* The SERVER must not inherit the browser's copy. The Rapport payload names the basemap the
  operator was looking at, and a keyed template in that payload would put the key into request
  bodies, application logs and the on-disk tile-cache filenames — three places it has no reason
  to be, on a machine that already holds the credential itself.

So the client sends the template UNKEYED (``withoutCartoBasemapKey``) and the server applies its
own credential here. Belt and braces: ``strip_key`` runs on the way in regardless, so an older
client that still sends a keyed template — or a station-authored custom layer with a key pasted
into it — cannot smuggle one into the cache key either.
"""

import re
from urllib.parse import quote

from .credentials import get as credential

#: The CARTO raster CDN, optionally on one of its a–d subdomains. Anything else is left alone:
#: swisstopo WMTS, a station's own tile server and a self-hosted style have no business
#: receiving this deployment's CARTO credential.
_CARTO_RASTER_HOST = re.compile(r"^https://(?:[a-d]\.)?basemaps\.cartocdn\.com/", re.IGNORECASE)


def is_carto(url: str) -> bool:
    """Is this a CARTO raster tile template?"""
    return bool(_CARTO_RASTER_HOST.match(url))


def strip_key(url: str) -> str:
    """Remove any ``key=`` parameter, leaving a template safe to log and to hash into a cache
    filename. Host-agnostic on purpose — a key does not become ours by sitting on another CDN,
    and a cache entry must never be split by a credential that can be rotated.

    ⚠️ Splits the query on its real separators rather than pattern-matching the URL: substring
    surgery is how ``monkey=1`` gets eaten. ``urlsplit`` is not used because the ``{z}/{x}/{y}``
    slots must survive verbatim.
    """
    base, sep, frag = url.partition("#")
    path, q, query = base.partition("?")
    if not q:
        return url
    kept = [p for p in query.split("&") if p and p != "key" and not p.startswith("key=")]
    return (f"{path}?{'&'.join(kept)}" if kept else path) + sep + frag


def with_key(url: str) -> str:
    """Apply THIS deployment's CARTO key to a CARTO raster template. A non-CARTO host, or no
    configured key, returns the URL untouched — an unkeyed CARTO request still answers, with the
    provider's watermarked tiles, so a station that has not set a key still gets a Rapport."""
    key = (credential("carto_api_key") or "").strip()
    if not key or not is_carto(url):
        return url
    return f"{url}{'&' if '?' in url else '?'}key={quote(key, safe='')}"


def for_fetch(url: str) -> tuple[str, str]:
    """``(fetch_url, cache_url)`` for one raster template.

    ``cache_url`` is the identity of the tile — no credential in it, so rotating the key does
    not orphan the cache and the cache filenames carry nothing secret. ``fetch_url`` is what
    actually goes to the CDN.
    """
    bare = strip_key(url)
    return with_key(bare), bare
